checks.d/mapreduce.py

```diff
@@ -0,0 +1,449 @@
+                        counter_name = counter.get('counter_name')
+
+                        if not counter_name:
+                            raise Exception('At least one "counter_name" should be specified in the list of "counters"')
+
+                        custom_metrics[job_name][counter_group_name].append(counter_name)
+
+        return custom_metrics
+
+    def _get_cluster_id(self, rm_address):
+        '''
+        Return the cluster ID, otherwise raise an exception
+        '''
+        info_json = self._rest_request_to_json(rm_address,
+            INFO_PATH,
+            service_name=YARN_SERVICE_CHECK)
+
+        return info_json.get('clusterInfo', {}).get('id')
+
+    def _get_running_app_ids(self, rm_address, **kwargs):
+        '''
+        Return a dictionary of {app_name: tracking_url} for the running MapReduce applications
+        '''
+        metrics_json = self._rest_request_to_json(rm_address,
+            YARN_APPS_PATH,
+            service_name=YARN_SERVICE_CHECK,
+            **kwargs)
+
+        running_apps = {}
+
+        if metrics_json['apps'] is not None:
+            if metrics_json['apps']['app'] is not None:
+
+                for app_json in metrics_json['apps']['app']:
+                    tracking_url = app_json['trackingUrl']
+                    app_name = app_json['name']
+
+                    if tracking_url is not None \
+                            and app_name is not None:
+
+                        running_apps[app_name] = tracking_url
+
+        return running_apps
+
+    def _mapreduce_job_metrics(self, running_apps, cluster_id):
+        '''
+        Get metrics for each MapReduce job.
+        Return a dictionary for each MapReduce job
+        {
+          job_name: {
+            'app_name': app_name,
+            'user_name': user_name,
+            'tracking_url': tracking_url
+        }
+        '''
+        running_jobs = {}
+
+        for app_name, tracking_url in running_apps.iteritems():
+
+            metrics_json = self._rest_request_to_json(tracking_url,
+                MAPREDUCE_JOBS_PATH,
+                service_name=MAPREDUCE_SERVICE_CHECK)
+
+            if metrics_json['jobs'] is not None:
+                if metrics_json['jobs']['job'] is not None:
+
+                    for job_json in metrics_json['jobs']['job']:
+                        job_id = job_json['id']
+                        job_name = job_json['name']
+                        user_name = job_json['user']
+
+                        if job_id is not None \
+                                and job_name is not None \
+                                and user_name is not None:
+
+                            running_jobs[str(job_name)] = {'app_name': str(app_name),
+                                                    'user_name': str(user_name),
+                                                    'tracking_url': self._join_url_dir(tracking_url, MAPREDUCE_JOBS_PATH, job_id)}
```

Comment:
I think that the `tracking_url` is specific to a `job_id`, not a `job_name`, so there should be multiple `tracking_url`s for a given `job_name`. Otherwise, when multiple jobs share the same `job_name`, we'd only retrieve the `MAPREDUCE_JOB_COUNTER_METRICS`, `MAPREDUCE_MAP_TASK_METRICS ` and `MAPREDUCE_REDUCE_TASK_METRICS` on one of the jobs that has that `job_name`, and the values on these metrics would only reflect that one job instead of all the jobs that share the same `job_name`.

For these metrics to be correct, we need to fetch the metrics for every job (even those that share the same `job_name`), and then compute the aggregates for each `job_name`.

Does that make sense?